lib/parsers/src/tool_calling/pythonic_parser.rs

```diff
@@ -0,0 +1,343 @@
+// SPDX-FileCopyrightText: Copyright (c) 2024-2025 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+use super::response::{CalledFunction, ToolCallResponse, ToolCallType};
+use rustpython_parser::{
+    Mode,
+    ast::{Constant, Expr, Mod},
+    parse,
+};
+use serde_json::{Number, Value, json};
+
+fn strip_text(message: &str) -> String {
+    // Remove unexpected python tags if any
+    message
+        .replace("<|python_start|>", "")
+        .replace("<|python_end|>", "")
+}
+
+fn get_regex_matches(message: &str) -> Vec<String> {
+    use regex::Regex;
+    let pattern = r"\[([a-zA-Z]+\w*\(([a-zA-Z]+\w*=.*?,\s*)*([a-zA-Z]+\w*=.*?\s?)?\),\s*)*([a-zA-Z]+\w*\(([a-zA-Z]+\w*=.*?,\s*)*([a-zA-Z]+\w*=.*?\s*)?\)\s*)+\]";
+    let re = Regex::new(pattern).unwrap();
+
+    let mut matches = Vec::new();
+    for cap in re.find_iter(message) {
+        matches.push(cap.as_str().to_string());
+    }
+    matches
+}
+
+pub fn parse_tool_calls(src: &str) -> anyhow::Result<Vec<ToolCallResponse>> {
+    let ast = parse(src, Mode::Expression, "<input>")?;
+
+    /*
+    AST: Expression(ModExpression {
+        range: (),
+        body: List(ExprList {
+            range: 0..25,
+            elts: [Call(...), Call(...)]
+            ctx: Load
+        })
+    })
+    */
+    let body = match ast {
+        Mod::Expression(mod_expr) => mod_expr.body,
+        _ => return Ok(vec![]),
+    };
+
+    let elts = match *body {
+        Expr::List(expr_list) => expr_list.elts,
+        _ => return Ok(vec![]),
+    };
+
+    let mut res = Vec::new();
+    for (idx, elt) in elts.iter().enumerate() {
+        let (func, keywords) = match elt {
+            Expr::Call(call) => (&call.func, &call.keywords),
+            _ => continue,
+        };
+
+        let name = match func.as_ref() {
+            Expr::Name(name) => name.id.clone(),
+            _ => continue,
+        };
+
+        let mut obj = serde_json::Map::new();
+        for keyword in keywords.iter() {
+            let Some(arg_ident) = keyword.arg.as_ref() else {
+                tracing::debug!(
+                    "Skipping **kwargs in pythonic tool call for function {}",
```

Comment:
When you test this locally, can you check if how noisy your debug calls are, and mostly remove them?